### PR TITLE
Registry tutorial - basic

### PR DESCRIPTION
**Pull request by ngrayluna**
W&B Registry tutorial, example using simple  PyTorch neural network for classifying animal classes.

The purpose of this notebook is to show how to use registry with the minimal amount of boiler plate code (so as to hopefully not drown out the W&B-specific portions of the notebook).

### CODE REVIEW

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 1]
```diff
@@ -0,0 +1,634 @@
+{
```
Comment (ijdoc) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <blockquote>From Noa's suggestion to use core registries instead:</blockquote><p>The goal of this notebook is to demonstrate how you and other members of your organization can use W&amp;B Registry to track, share, and use dataset and model artifacts in your machine learning workflows. By the end of this notebook, you will know how to use W&amp;B to:</p><ol><li>Use a&nbsp;<a href="https://docs.wandb.ai/guides/registry/registry_types#core-registry" target="_blank">core registry</a></li><li>Create&nbsp;<a href="https://docs.wandb.ai/guides/registry/create_collection" target="_blank">collections</a>&nbsp;within a registry</li><li>Make dataset and model artifacts available to other members of your organization, and</li><li>Download artifacts from the registry for inference</li></ol><p>To achieve this, we will train a neural network to identify animal classes, as an example.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (ijdoc) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p>This can be on a different or follow-up notebook as per Noa's suggestion to use the core registries + collections</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (ijdoc) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p>Since we are logging tensors, there is an opportunity to call this collection "zoo-dataset-tensors" or similar, within the core Datasets registry, which would differ, for example, from "zoo-dataset-tensors-split" created in the next section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (ijdoc) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p>There is an oportunity to use wandb to track this split (using the function arguments as config), and then:</p><ul><li>Make a note that, although optional, tracking this split can simplify reproducibility later on</li><li>Make a note (with a doc link) explaining that doing this will also reflect the split in the lineage graph, which improves observability</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

**Review by tcapelle — APPROVED** [review]

**Comment by tcapelle** [comment]
Added a name, a tracker, colab badge and clean up =)

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 2]
```diff
@@ -0,0 +1,984 @@
...
```
Comment (tcapelle) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <pre class="ql-syntax" spellcheck="false">[class_labels.get(x) for x in predicted.numpy()]
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (tcapelle) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p>what about this commented pieces?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='59'/>

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 3]
```diff
@@ -0,0 +1,984 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "507ab87f",
+   "metadata": {},
+   "source": [
+    "The goal of this notebook is to demonstrate how you and other members of your organization can use W&B Registry to track, share, and use dataset and model artifacts in your machine learning workflows. By the end of this notebook, you will know how to use W&B to:\n",
+    "\n",
+    "1. Create [collections](https://docs.wandb.ai/guides/registry/create_collection) within the [Dataset and Model registry](https://docs.wandb.ai/guides/registry/registry_types#core-registry).\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:36]: This should just be Registry no? We are trying to make it abstract to more than just models and artifacts?

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 4]
```diff
@@ -0,0 +1,984 @@
...
+    "2. Make dataset and model artifacts available to other members of your organization, and\n",
+    "3. Download our trained model and dataset artifacts from the registry for inference\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:38]: nit: "Download your trained model..."

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 5]
```diff
@@ -0,0 +1,984 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b87ee0c6-4e88-4c77-9cf5-dde109bb7ebd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!pip install wandb torch ucimlrepo scikit-learn pandas"
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:58]: Should be un-commented by default IMO, since a new colab runtime won't have these (likely where most notebooks are ran) and if running locally it will just display already installed packages

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 6]
```diff
@@ -0,0 +1,984 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "742c237c-1e8d-4c2e-8ac4-8221146d876e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "PROJECT = \"zoo_experiment\"\n",
+    "TEAM_ENTITY = \"smle-reg-team-2\"\n",
+    "\n",
+    "run = wandb.init(\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:259]: It'd be cleaner to ask the user to input a WANDB_API_KEY upstream to not block them in the middle of running the notebook interactively?

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 7]
```diff
@@ -0,0 +1,984 @@
...
+    "    entity=TEAM_ENTITY,\n",
+    "    project=PROJECT,\n",
+    "    job_type=\"publish_dataset\"\n",
+    ")\n",
+    "\n",
+    "artifact = wandb.Artifact(\n",
+    "    name=\"zoo_dataset\",\n",
+    "    type=\"dataset\", \n",
+    "    description=\"Processed dataset and labels.\"\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:268]: nit: "datasets", we want the end user to know this artifact will contain multiple files.

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 8]
```diff
@@ -0,0 +1,984 @@
...
+    ")\n",
+    "\n",
+    "artifact.add_file(local_path=\"zoo_dataset.pt\", name=\"zoo_dataset\")\n",
+    "artifact.add_file(local_path=\"zoo_labels.pt\", name=\"zoo_labels\")\n",
+    "\n",
+    "run.link_artifact(artifact=artifact, target_path=dataset_target_path)\n",
+    "\n",
+    "run.finish()"
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:276]: given we are doing all run actions in one cell, it is potentially cleaner to use a context manager here? `with wandb.init() as run: ...  etc`

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 9]
```diff
@@ -0,0 +1,984 @@
...
+    "artifact.add_file(local_path=\"zoo_dataset_X_test.pt\", name=\"zoo_dataset_X_test\")\n",
+    "artifact.add_file(local_path=\"zoo_labels_y_test.pt\", name=\"zoo_labels_y_test\")\n",
+    "\n",
+    "REGISTRY_NAME = \"Dataset\"\n",
+    "COLLECTION_NAME = \"zoo-dataset-tensors-split\"\n",
+    "target_dataset_path=f\"{ORG_NAME}/wandb-registry-{REGISTRY_NAME}/{COLLECTION_NAME}\"\n",
+    "\n",
+    "run.link_artifact(artifact=artifact, target_path=target_dataset_path)\n",
+    "\n",
+    "run.finish()"
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:368]: Potentially a good idea to use a context manager here too, it helps visualize how little code they actually have to add in order to log to W&B.

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 10]
```diff
@@ -0,0 +1,984 @@
...
+    "            name=model_artifact_name,\n",
+    "            type=\"model\",\n",
+    "            metadata={\n",
+    "                \"num_classes\": 7,\n",
+    "                \"model_type\": wandb.config[\"model_type\"]\n",
+    "            }\n",
+    "        )\n",
+    "        \n",
+    "        # Add artifact file\n",
+    "        artifact.add_file(PATH)\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:540]: Shouldn't this be outside the for loop so we only save the best model? just save a ref to the best model inside the for loop, otherwise we are saving a bunch of iterations that aren't necessary, or are we doing this on purpose to showcase versioning?

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 11]
```diff
@@ -0,0 +1,984 @@
...
+    "        artifact.save()\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:541]: Use run.log_artifact() to emphasize the association with this run?

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 12]
```diff
@@ -0,0 +1,984 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8d4ed37c-f488-4cd8-8797-fdeb6002148e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "run = wandb.init(entity=TEAM_ENTITY, project=PROJECT)\n",
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:656]: Context manager?

File: colabs/wandb_registry/zoo_wandb.ipynb [hunk 13]
```diff
@@ -0,0 +1,984 @@
...
...aMnH2bLQg+TJKFGCBKHVBKlXEHAZgXv37tPZs+f5w5ExM6YP00bfRoMsAOeKMRaigYmnb5dfPylAEBAEBAFBwIMQgLG375zgPBUsmDdc6U+d1RHg8UEnICQwD3qRpSlhFoESRQtTzhzZaemK1XTp0hW6cu0/evHiBaVNk4rSp0tLn/C9fHk/DbP9l44JAiGJgNYRIOoI5gRhVdypI9DOYoxlgggGwT0RQUAQEAQEAUEgKAhAZy8iCAgCgoAgIAgIAoKAIOA4Am/fvg1/BDDAY4wCBpIWPsEtqFNEEPBWBHbu2qM2FrJmyUzJkib11m5Iuz0EgSyZM1Hnju1cak3evLldyu8tmUHeSs0G1a5IosQJXckueQUBQUAQEATCAALnbl62kLjQnc9zFXeYxAWv4NrAC3nXHuJIYJ+lwamIIGBBoHrVioSPSNhG4Mc2LcJ2B6V34QIBGHSB+KUNvrG5HBYNvBRZy8/4yhXil/Gl0J6+jUSwM1d8o4IZ08m5ICAICAKCgCDgDQgg8hfmAfjs2rUvzBPDjc/EWkeQIUlqCkokL60jQEQxRAILSl5jO+RcEBAE7CMQP15calSvjv1EclcQEARcRgA6AcwF9JwAcwToDsKa6PU7InpnShXdLUQtkL2ypIlOK7bdU1HDQQITAlhYe3OkP4KAICAIhBwCCRKEvfE35NCTmgQBQUAQEAQEAUEgvCEAEpiKAPb+/XtCSLDwIIgABkJWSBC/dF04iggC3orA02fPqGrlilS0SEFv7YK024MQAKGp3jdfe1CLPLcpGDtk/PDc5yMtEwQEAUHAWxCAUZZRMtxI7TABDPlAAoOHcBh2oSx8cE1EEBAEBAFBQBDwRgRgzAUBCSysG3ihn+42vgIRTMupy0IA01jIURAQBAQBQcD7EIDBtzESGHqg5wne1xvbLUZkb6zntVjrCPA9qCQukMDazRmoilxzaEuQ8+u2yFEQEAQEAUFAEAhtBHTkL8wJtI4gLEYM1+t3d5G/9HODjqBysY/o1O/X1CU4jTHqDXQ6OQoCgoAgIAgIArYQwPgLCYsEbFt9luuCgCAgCAgCgoAgIAi4goCPjw+9e/eOfMIT+UsD1qpFU8JnwiTf6F979+3Xt9x2BMlMDPfdBqcUFIoIVKzwGeEjIggIAoKAICAICAKCgCDg+QjAEAsGWGYCwy9nyFswBtMGYxIF
```
Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:673]: This supposed to be an image? Doesn't render for me.

Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:763]: Context manager?

Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:797]: Would be helpful to explain how / why you arrived at this path from .download()

Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:834]: Context manager?

Comment (abraham-leal) [inline colabs/wandb_registry/zoo_wandb.ipynb:964]: Feels unfinished?

**Review by abraham-leal — APPROVED** [review]
Overall awesome, lots of stylistic suggestions for readability. Thank you!

File: colabs/wandb_registry/zoo_wandb.ipynb (hunk 2 above)
Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p>-Sorry for the product marketing nitpickiness as there are tons of discrepancies across our docs and sites regarding capitalization and naming for the Registry feature. In this context, "Dataset" and "Model" should not be capitalized when discussing the dataset registry and the model registry, both of which are part of Registry.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>There are two ways to get the target path, "interatively" -&gt; "interactively"</li><li>Not sure it's even necessary to add the commented out interactive method of getting the target path here, but will leave that up to the author. </li><li>The target path of a "collectin" consists of three parts -&gt; "collection"</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>Is it possible to use an ORG_NAME not associated with the bug bash? Actually wondering if we need to just use &lt;INSERT YOUR ORG_NAME HERE&gt; syntax as this will only actually work if you have a key associated with that org (please correct me if I am wrong about this).</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>created the artifact object with&nbsp;<code>wandb.Aritfact()</code>) -&gt; "wandb.Artifact()"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>Is it possible to use an ORG_NAME not associated with the bug bash? Actually wondering if we need to just use &lt;INSERT YOUR ORG_NAME HERE&gt; syntax as this will only actually work if you have a key associated with that org (please correct me if I am wrong about this).</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>The preceding code blocks links -&gt; "block"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>View lineage map of registired model -&gt; "registered"</li><li>The lineage shows which artifcats were -&gt; "artifacts"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <p> You want to retrieve the model and dataset artifact was pushed to the registry -&gt; "artifact pushed to the registry"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>Perhaps change the ORG_NAME to &lt;ENTER YOUR ORG_NAME HERE&gt; (or something along those lines)</li><li>VERESION = 0 -&gt; "VERSION"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (rratshin-wandb) [inline colabs/wandb_registry/zoo_wandb.ipynb:1]: <ul><li>Will there be any conclusion text here to wrap-up the lesson?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/wandb/examples/pull/560/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

File: colabs/wandb_registry/zoo_wandb.ipynb (hunk 10 above)
Comment (ngrayluna) [reply colabs/wandb_registry/zoo_wandb.ipynb:540]: Yeah, the idea was to show versioning. But, thinking about it more, showing it this way might convolute the example.

In a real world setting, I suspect versioning is more intuitive to see, use as they are messing about with hyperparameters, modifying the neural network itself, etc. 

TL;DR Showing versioning here as a byproduct of checkpointing might not be intuitive, useful for showing why, how to use versioning.

File: colabs/wandb_registry/zoo_wandb.ipynb (hunk 8 above)
Comment (ngrayluna) [reply colabs/wandb_registry/zoo_wandb.ipynb:276]: Good point. It's tricky. From Python POV it totally makes sense. On one end, some folks find context manager confusing. Chatted with Morgan some time ago and we mentioned about creating a "best practices" guide...but we never got around to it. 

For simplicity, I'll keep it as is (without the context manager) but am open to updating it in the future if that's the path we collectively decide makes sense from a teaching point of view.

**Comment by ngrayluna** [comment]
Pushed edits, things to resolve (see Slack thread for answers)
1. Add action that replaces org variable strings w/ actual org name for CI 
2. Once tags ship (SDK) add that to notebook

Note: Might need to update this notebook if we hide org names.
